src/distributions/uniform.rs

```diff
@@ -494,9 +495,64 @@ macro_rules! uniform_float_impl {
 uniform_float_impl! { f32, 32 - 23, next_u32 }
 uniform_float_impl! { f64, 64 - 52, next_u64 }
 
+/// Implementation of `UniformImpl` for `Duration`.
+/// 
+/// Unless you are implementing `UniformImpl` for your own types, this type should
+/// not be used directly, use `Uniform` instead.
+#[derive(Clone, Copy, Debug)]
+pub struct UniformDuration {
+    size: Duration,
+    offset: Duration,
+}
+
+impl SampleUniform for Duration {
+    type Impl = UniformDuration;
+}
+
+impl UniformImpl for UniformDuration {
+    type X = Duration;
+
+    #[inline]
+    fn new(low: Duration, high: Duration) -> UniformDuration {
+        UniformDuration::new_inclusive(low, high - Duration::new(0, 1))
+    }
+
+    #[inline]
+    fn new_inclusive(low: Duration, high: Duration) -> UniformDuration {
+        UniformDuration {
+            size: high - low,
+            offset: low,
+        }
+    }
+
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Duration {
+        // the basic approach here is to "round up" to the next second and
+        // uniformly sample from that, retrying if we get a value outside of
+        // the range. As long as our range is larger than 1 second this has a
+        // > 50% chance of succeeding. We special case subsecond sizes because
+        // they're easy and to avoid low success rates of the normal approach.
+        let d = if self.size < Duration::from_secs(1) {
+            let nanos = Uniform::new_inclusive(0, self.size.subsec_nanos()).sample(rng);
+            Duration::new(0, nanos)
+        } else {
+            loop {
```

Comment:
I couldn't think of an approach like that which avoided a bias in the nanoseconds, since the sampling ranges differ based on if you've selected the maximum number of seconds or not.

Reply:
Oooh, were you thinking of something like this?

```rust
let sec = Uniform::new_inclusive(0, self.size.as_secs());
let nsec = if sec == self.size.as_secs() {
    let mut nsec_gen = Uniform::new(0, 1_000_000_000);
    loop {
        let nsec = nsec_gen.sample(rng);
        if nsec <= self.size.subsec_nanos() {
            break nsec;
        }
    }
} else {
    rng.gen_range(0, 1_000_000_000)
};

self.offset + Duration::new(sec, nsec)
```

That should work.

Reply:
Quick work :smile:.

Oops, didn't think that through fully.
What about:
1. generate seconds
2. if generated second < max, generate nanos in the (0, 1_000_000_000) range.
3. if generated second == max or seconds == 0, generate nanos in custom range.

We kind of like `Uniform` to be as exact as possible...

And you could move the cached range to generate seconds up to `new` or `new_inclusive` if you want.

Reply:
Oh no actually this has the nonuniformity problems I mentioned.

Reply:
The alternative would be leaning on the u128 implementation but that wouldn't be universally applicable.

Reply:
Just saw your comment just after posting. Yes, that is in the direction, but you don't need a loop (just realizing).

And it may be worth it to skip generating seconds if it is 0.

Reply:
The current approach is exactly uniform, but the no-loop approach has uniformity issues. Imagine if we wanted to sample from 0 to 1.1 seconds. We'll pick 0 or 1 full seconds uniformly, but that means we'll pick 0-1 seconds 50% of the time, and 1-1.1 seconds 50% of the time.

Reply:
I think the steps I wrote above don't have uniformity problems.

I should write it down a bit better, but these comments are from a tabled and it is late for me. Do you want to play with it some more? Otherwise I can give it a try tomorrow.

Reply:
Let's circle back tomorrow!